### PR TITLE
Update MultiheadAttention module support key/value with different number of features and allow static key/value

### PR DESCRIPTION
**Pull request by zhangguanheng66**
The changes include:

1. Allow key/value to have different number of features with query. It supports the case when key and value have different feature dimensions.
2. Support three separate proj_weight, in addition to a single in_proj_weight. The proj_weight of key and value may have different dimension with that of query so three separate proj_weights are necessary. In case that key and value have same dimension as query, it is preferred to use a single large proj_weight for performance reason. However, it should be noted that using a single large weight or three separate weights is a size-dependent decision.
3. Give an option to use static k and v in the multihead_attn operator (see saved_k and saved_v). Those static key/value tensors can now be re-used when training the model.
4. Add more test cases to cover the arguments.

Note: current users should not be affected by the changes.

### CODE REVIEW

File: torch/nn/modules/activation.py [hunk 1]
```diff
@@ -691,27 +691,39 @@ class MultiheadAttention(Module):
...
+        query, key, and value have the same number of features.
+
     Examples::
 
         >>> multihead_attn = nn.MultiheadAttention(embed_dim, num_heads)
         >>> attn_output, attn_output_weights = multihead_attn(query, key, value)
     """
 
-    def __init__(self, embed_dim, num_heads, dropout=0., bias=True, add_bias_kv=False, add_zero_attn=False):
+    def __init__(self, embed_dim, num_heads, kdim=None, vdim=None, dropout=0., bias=True, add_bias_kv=False, add_zero_attn=False):
```
Comment (soumith) [inline torch/nn/modules/activation.py]: the added args have to go last, or else backward compatibility breaks

File: torch/nn/modules/activation.py [hunk 2]
```diff
@@ -691,27 +691,39 @@ class MultiheadAttention(Module):
...
+        self.qkv_same_dim = self.kdim == embed_dim and self.vdim == embed_dim
+
         self.num_heads = num_heads
         self.dropout = dropout
         self.head_dim = embed_dim // num_heads
         assert self.head_dim * num_heads == self.embed_dim, "embed_dim must be divisible by num_heads"
 
-        self.in_proj_weight = Parameter(torch.empty(3 * embed_dim, embed_dim))
+        self.q_proj_weight = Parameter(torch.Tensor(embed_dim, embed_dim))
+        self.k_proj_weight = Parameter(torch.Tensor(embed_dim, self.kdim))
```
Comment (soumith) [inline torch/nn/modules/activation.py]: when you load older models / older models' state-dicts, the model will break / compute wrong result. what are you going to do about that?

you can do what batchnorm did. see https://github.com/pytorch/pytorch/blob/master/torch/nn/modules/batchnorm.py#L16 and look at `_version = 2`,  and https://github.com/pytorch/pytorch/blob/master/torch/nn/modules/batchnorm.py#L90-L103

**Review by soumith — CHANGES_REQUESTED** [review]

File: torch/nn/modules/activation.py (hunk 2 above)
Comment (zhangguanheng66) [reply torch/nn/modules/activation.py]: Thanks for pointing this out.

add _load_from_state_dict() function to detect in_proj_weight. If in_proj_weight exists in state_dict, it will split in_proj_weight into three separate weights. I assume users know how to use state_dict() and load_state_dict() to resolve this type of version conflict.

Test on the word language model
1. load the model based by the old module, which has in_proj_weight. 
`model = torch.load("old.pt")`
2. generate state_dict. 
`state_dict = model.state_dict()`
3. create a model based on the new module, which has three separate weights. 
`new_model = new_module()`
4. map state on the new model. 
`new_model.load_state_dict(state_dict)`

File: torch/nn/functional.py [hunk 3]
```diff
@@ -3113,6 +3119,9 @@ def multi_head_attention_forward(query,                  # type: Tensor
             be ignored by the attention.
         need_weights: output attn_output_weights.
         attn_mask: mask that prevents attention to certain positions.
+        use_chunk_proj_weight: use in_proj_weight insteady of q_proj_weight, k_proj_weight, v_proj_weight.
```
Comment (cpuhrsch) [inline torch/nn/functional.py]: Nit: "instead"

Also, this means that passing the flags for q, k, v weights is now the default? That's breaking backwards compatibility. Is that worth it?

File: torch/nn/functional.py [hunk 4]
```diff
@@ -3125,6 +3134,10 @@ def multi_head_attention_forward(query,                  # type: Tensor
           the embedding dimension.
         - key_padding_mask: :math:`(N, S)`, ByteTensor, where N is the batch size, S is the source sequence length.
         - attn_mask: :math:`(L, S)` where L is the target sequence length, S is the source sequence length.
+        - saved_k: :math:`(N*num_heads, S, E/num_heads)`, where S is the source sequence length,
```
Comment (cpuhrsch) [inline torch/nn/functional.py]: Why is it called "saved"?

File: torch/nn/modules/activation.py [hunk 5]
```diff
@@ -772,11 +789,26 @@ def forward(self, query, key, value, key_padding_mask=None,
         - attn_output_weights: :math:`(N, L, S)` where N is the batch size,
           L is the target sequence length, S is the source sequence length.
         """
-        return F.multi_head_attention_forward(
-            query, key, value, self.embed_dim, self.num_heads,
-            self.in_proj_weight, self.in_proj_bias, self.bias_k, self.bias_v, self.add_zero_attn,
-            self.dropout, self.out_proj.weight, self.out_proj.bias, training=self.training,
-            key_padding_mask=key_padding_mask, need_weights=need_weights, attn_mask=attn_mask)
+        if hasattr(self, '_qkv_same_embed_dim') and self._qkv_same_embed_dim is False:
```
Comment (cpuhrsch) [inline torch/nn/modules/activation.py]: Why do you check it using an attribute?

File: torch/nn/functional.py [hunk 6]
```diff
@@ -3145,73 +3158,94 @@ def multi_head_attention_forward(query,                  # type: Tensor
...
-        k = linear(key, _w, _b)
-
-        _b = in_proj_bias
-        _start = embed_dim * 2
-        _end = None
-        _w = in_proj_weight[_start:, :]
-        if _b is not None:
-            _b = _b[_start:]
-        v = linear(value, _w, _b)
+        assert q_proj_weight is not None
```
Comment (cpuhrsch) [inline torch/nn/functional.py]: Add a message to be printed if the assert fails or catch it higher up.

File: torch/nn/modules/activation.py (hunk 5 above)
Comment (zhangguanheng66) [reply torch/nn/modules/activation.py]: This will resolve the backward capability issue.

File: torch/nn/functional.py (hunk 4 above)
Comment (zhangguanheng66) [reply torch/nn/functional.py]: we can use "static_k, static_v".

File: torch/nn/functional.py (hunk 3 above)
Comment (ngimel) [reply torch/nn/functional.py]: It preserves backwards compatiblity (use_chunk_proj_weight=True will use in_proj_weight, like before) but the name is confusing.

File: torch/nn/functional.py [hunk 7]
```diff
@@ -3145,73 +3158,94 @@ def multi_head_attention_forward(query,                  # type: Tensor
...
+        q_embed_dim = query.size(-1)
+        assert tuple(q_proj_weight.size()) == (embed_dim, q_embed_dim)
+
+        assert k_proj_weight is not None
+        k_embed_dim = key.size(-1)
+        assert tuple(k_proj_weight.size()) == (embed_dim, k_embed_dim)
+
+        assert v_proj_weight is not None
+        q = linear(query, q_proj_weight, in_proj_bias[0:embed_dim])
```
Comment (ngimel) [inline torch/nn/functional.py]: can `in_proj_bias` be `None`? Also, it seems like this case can share most of the code with the case where neither qkv_same nor kv_same is true, but assigning q_proj_weight and the like to the slices of the in_proj_weight matrix.

File: torch/nn/functional.py [hunk 8]
```diff
@@ -3145,73 +3158,94 @@ def multi_head_attention_forward(query,                  # type: Tensor
...
+        k = linear(key, k_proj_weight, in_proj_bias[embed_dim:(embed_dim * 2)])
+        v = linear(value, v_proj_weight, in_proj_bias[(embed_dim * 2):])
+        v_embed_dim = value.size(-1)
+        assert tuple(v_proj_weight.size()) == (embed_dim, v_embed_dim)
+
     q *= scaling
```
Comment (ngimel) [inline torch/nn/functional.py]: Forgot to mention it earlier, but this in-place multiplication really hurts performance if q is non-contiguous (which it usually is).

File: torch/nn/functional.py (hunk 7 above)
Comment (zhangguanheng66) [reply torch/nn/functional.py]: good catch for the in_proj_bias

File: torch/nn/functional.py (hunk 8 above)
Comment (zhangguanheng66) [reply torch/nn/functional.py]: Thanks. We will try to optimize this.

Comment (ngimel) [reply torch/nn/functional.py]: Just do q=q*scaling, iirc

File: torch/nn/functional.py (hunk 6 above)
Comment (zhangguanheng66) [reply torch/nn/functional.py]: the jit team instructed me to convert optional tensor q_proj_weight to non-optional tensor q_proj_weight_non_opt. The casting will throw error message if q_proj_weight is None.

**Comment by zhangguanheng66** [comment]
fix https://github.com/pytorch/pytorch/issues/21518

File: torch/nn/functional.py (hunk 3 above)
Comment (zhangguanheng66) [reply torch/nn/functional.py]: I have added more docs for use_chunk_proj_weight

File: torch/nn/functional.py [hunk 9]
```diff
@@ -3091,23 +3091,29 @@ def _pad_circular(input, padding):
...
+                                 out_proj_bias,                   # type: Tensor
+                                 training=True,                   # type: bool
+                                 key_padding_mask=None,           # type: Optional[Tensor]
+                                 need_weights=True,               # type: bool
+                                 attn_mask=None,                  # type: Optional[Tensor]
+                                 use_separate_proj_weight=False,  # type: bool
+                                 q_proj_weight=None,              # type: Optional[Tensor]
+                                 k_proj_weight=None,              # type: Optional[Tensor]
+                                 v_proj_weight=None,              # type: Optional[Tensor]
+                                 static_k=None,                   # type: Optional[Tensor]
```
Comment (cpuhrsch) [inline torch/nn/functional.py]: I think we had talked about maybe moving this into a separate function by splitting multi_head_attention_forward out into two, one part that does the projections and the other that consumes them. Users that want to do their own projections can then call the latter.

File: torch/nn/functional.py (hunk 8 above)
Comment (cpuhrsch) [reply torch/nn/functional.py]: @ngimel - on an unrelated note, and maybe you don't have context on this, are we well-advised to bind the cudnn implementation of multiheadattention at this point?

File: torch/nn/functional.py (hunk 9 above)
Comment (zhangguanheng66) [reply torch/nn/functional.py]: But it may not really help the layout of proj_weights issues. We still have to maintain the two layouts at the same time, right?

File: torch/nn/functional.py (hunk 8 above)
Comment (ngimel) [reply torch/nn/functional.py]: At this point it is not ready for prime-time yet.

File: torch/nn/functional.py (hunk 9 above)
Comment (ngimel) [reply torch/nn/functional.py]: If you (or user) pass separate `q_proj_weight`, `k_proj_weight`, `v_proj_weight` the layout is pretty flexible, and 2 layouts are not really necessary. But this implies that (worst case) 3 gemms will be called instead of 1, which may or may not hurt performance.

Comment (zhangguanheng66) [reply torch/nn/functional.py]: @ngimel yeap. @cpuhrsch has this performance concern so we keep the single proj_weight option here. Only use the three gemms option when key/value have different embed dimension with query.

File: test/test_nn.py [hunk 10]
```diff
@@ -3410,63 +3405,89 @@ def _multihead_attn_test_helper(add_key_padding_mask, add_bias_kv=False, add_zer
...
+                Q_fc = _fc(Q, q_proj_weight, multihead_attn_module.in_proj_bias[:d_model])
+                K_fc = _fc(K, k_proj_weight, multihead_attn_module.in_proj_bias[d_model:(d_model * 2)])
+                V_fc = _fc(V, v_proj_weight, multihead_attn_module.in_proj_bias[(d_model * 2):])
 
                 if add_bias_kv:
                     K_fc = np.concatenate((K_fc, np.repeat(bias_k, K_fc.shape[0], axis=0)), axis=1)
                     V_fc = np.concatenate((V_fc, np.repeat(bias_v, V_fc.shape[0], axis=0)), axis=1)
-                    attn_mask = np.concatenate((attn_mask, np.ones([1, 1])), axis=1)
+                    if attn_mask is not None:
+                        attn_mask = np.concatenate((attn_mask, np.ones([1, 1])), axis=1)
```
Comment (ngimel) [inline test/test_nn.py]: What are these `np` functions everywhere and how will they work with cuda tensors?

Comment (zhangguanheng66) [reply test/test_nn.py]: they are used to re-write the multiheadattention function and verify the results of nn.MulitheadAttention module.

File: torch/nn/functional.py (hunk 9 above)
Comment (ngimel) [reply torch/nn/functional.py]: fwiw, we used 3 separate gemms in mlperf submission, because then .contiguous() calls here https://github.com/pytorch/pytorch/blob/6d1f0daa25a3cde30dd4f772f4157e6bdde6a720/torch/nn/functional.py#L3276-L3280 become no-ops, so it compensates for gemms potentially being less efficient. Overall which approach is preferable is size-dependent. 
Alternatively, pointers and strides can be checked for projection weights and if they point to a contiguous matrix a single gemm can be called.

File: torch/nn/modules/activation.py (hunk 5 above)
Comment (cpuhrsch) [reply torch/nn/modules/activation.py]: I don't think this is necessary because we're updating the functional together with the module.

File: torch/nn/functional.py [hunk 11]
```diff
@@ -3159,73 +3176,99 @@ def multi_head_attention_forward(query,                  # type: Tensor
...
+                _b = in_proj_bias
+                _start = embed_dim
+                _end = None
+                _w = in_proj_weight[_start:, :]
+                if _b is not None:
+                    _b = _b[_start:]
+                k, v = linear(key, _w, _b).chunk(2, dim=-1)
+
         else:
+            _b = in_proj_bias
```
Comment (cpuhrsch) [inline torch/nn/functional.py]: Is this is getting quite complicated it could be good to add some comments.

File: torch/nn/modules/activation.py (hunk 5 above)
Comment (zhangguanheng66) [reply torch/nn/modules/activation.py]: some models trained by the old multiheadattention modules will break if there is not such a check, if my understanding is correct. Because they don't have q/k/v_proj_weight attributes.
